staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1019 @@
+	}
+	if err != nil && k8serrors.IsNotFound(err) {
+		// If we're here the remote attachment has been deleted.
+		deletedIfc, ifcFound, err := ca.deleteRemoteIfc(ifcKey)
+		if ifcFound {
+			glog.V(3).Infof("Deleted remote Network Interface %#+v\n", deletedIfc)
+		}
+		return err
+	}
+	return ca.createOrUpdateRemoteAttIfc(remoteAtt)
+}
+
+func (ca *ConnectionAgent) releaseLocalAttResources(attRef attQueueRef) error {
+	ca.updateVNStateForDeletedLocalAtt(attRef)
+	deletedIfc, ifcFound, err := ca.deleteLocalIfc(fromAttRefToVNIAndNsn(attRef))
+	if err == nil && ifcFound {
+		glog.V(3).Infof("Deleted local Network Interface %#+v\n", deletedIfc)
+	}
+	return err
+}
+
+// TODO break this method body into smaller methods, possibly shared with createOrUpdateRemoteAttIfc. Refactor it, make it more readable.
+func (ca *ConnectionAgent) processExistingLocalAtt(localAtt *netv1a1.NetworkAttachment) error {
+	ca.updateVNStateForExistingLocalAtt(localAtt)
+	ifcKey := fromAttToVNIAndNsn(localAtt)
+	oldIfc, ifcFound := ca.getLocalIfc(ifcKey)
+	var ifcNeedsUpdate bool
+	if ifcFound {
+		ifcNeedsUpdate = ca.ifcUpdateIsNeeded(oldIfc, localAtt)
+	}
+	var err1 error
+	var ifcWasUpdated bool
+	if ifcNeedsUpdate {
+		// Delete the old interface
+		// TODO think about adding an "UpdateIfc" method to the network fabric
+		err1 = ca.netFabric.DeleteLocalIfc(oldIfc)
+		if err1 != nil {
+			return err1
+		}
+		ca.unsetLocalIfc(ifcKey)
+		ifcWasUpdated = true
+	}
+	var newIfc netfabric.NetworkInterface
+	if !ifcFound || ifcNeedsUpdate {
+		// Create a new interface
+		// TODO could not think of a better way to generate a name. Think this through.
+		newIfc.Name = fmt.Sprintf("%d-%s/%s", localAtt.Spec.VNI, localAtt.Namespace, localAtt.Name)
+		newIfc.VNI = localAtt.Spec.VNI
+		newIfc.HostIP = gonet.ParseIP(ca.hostIP)
+		guestIP := gonet.ParseIP(localAtt.Status.IPv4)
+		newIfc.GuestIP = guestIP
+		newIfc.GuestMAC = generateMACAddr(localAtt.Spec.VNI, guestIP)
+		err1 = ca.netFabric.CreateLocalIfc(newIfc)
+		if err1 == nil {
+			ca.setLocalIfc(newIfc, ifcKey)
+			ifcWasUpdated = true
+		}
+	}
+	// Do a conditional update on the status of the NetworkAttachment
+	var err2 error
+	if localAtt.Status.HostIP != ca.hostIP || (ifcWasUpdated && (newIfc.Name != oldIfc.Name)) {
+		var ifcNameForStatus string
+		if ifcWasUpdated {
+			ifcNameForStatus = newIfc.Name
+		} else {
+			ifcNameForStatus = oldIfc.Name
+		}
+		var updatedAtt *netv1a1.NetworkAttachment
+		updatedAtt, err2 = ca.setHostIPAndIfcNameInAttStatus(localAtt, ifcNameForStatus)
+		if err2 == nil {
+			glog.V(4).Infof("Set host IP %s and interface name %s in local NetworkAttachment %s/%s status",
+				updatedAtt.Status.HostIP,
+				updatedAtt.Status.IfcName,
+				updatedAtt.Namespace,
+				updatedAtt.Name)
+		}
+	}
+	_, errToReturn := aggregateErrors("\n\t", err1, err2)
+	return errToReturn
+}
+
+// TODO the algorithm in this method is terribly complicated. Try to simplify it and review it to make sure it's correct
```

Comment:
You can simplify this and related functions, and remove the need for a mutex in `vnState`.  The key idea is that what is done while holding the lock bundles together a few critical operations that do not wait for anything.  See how the `interestsMutex` at https://github.ibm.com/k8s4g/hermes-iobricks-network-workspace/blob/8f7169a53946b9f3d08a4556c8931fa50d980da4/src/github.ibm.com/genctl/network-controllers/pkg/connection-agent/agent.go#L659 applies only to some maps and sets.  See how `setVPCInterest` at https://github.ibm.com/k8s4g/hermes-iobricks-network-workspace/blob/8f7169a53946b9f3d08a4556c8931fa50d980da4/src/github.ibm.com/genctl/network-controllers/pkg/connection-agent/agent.go#L1976 creates the new informers if needed, and updates the relevant tracking set of names.  See how `removeVPCInterest` at https://github.ibm.com/k8s4g/hermes-iobricks-network-workspace/blob/8f7169a53946b9f3d08a4556c8931fa50d980da4/src/github.ibm.com/genctl/network-controllers/pkg/connection-agent/agent.go#L2141 updates the relevant maps and sets, and shuts down the informer(s) if necessary, and computes the lists of names to enqueue, while holding the `interestsMutex`.